File: abc383b/abc383b.hs
{-# LANGUAGE OverloadedStrings #-}

import Control.Monad (replicateM)
import Data.Vector.Unboxed (Vector, (!), (//))
import qualified Data.Vector.Unboxed as V

main :: IO ()
main = do
  [h, w, d] <- f
  solve h w d <$> replicateM h getLine >>= print
  where
    f = map read <$> words <$> getLine

dist (xi, yi) (xj, yj) =
  abs (xi - xj) + abs (yi - yj)

solve :: Int -> Int -> Int -> [String] -> Int
solve h w d ss = f mp 0 1 0 0 0
  where
    mp = V.fromList $ concat ss
    f mp i j k c mc
      | i >= h * w = mc
      | j >= h * w = f mp (i + 1) (i + 2) 0 c mc
      | k >= h * w = f mp i (j + 1) 0 0 (max c mc)
      | mp ! i == '.' &&
        mp ! j == '.' &&
        mp ! k == '.' && (dist (hi, wi) (hk, wk) <= d || dist (hj, wj) (hk, wk) <= d) = f mp i j (k + 1) (c + 1) mc
      | otherwise = f mp i j (k + 1) c mc
      where
        (hi, wi) = i `divMod` w
        (hj, wj) = j `divMod` w
        (hk, wk) = k `divMod` w

File: abc383b/abc383b.ml
open Printf
open Scanf

let dist (x0, y0) (x1, y1) =
  abs (x0 - x1) + abs (y0 - y1)
                    
let solve h w d a =
  let n = h * w in
  let rec loop i j k c mc =
    if i >= n then mc
    else if j >= n then loop (i + 1) (i + 2) 0 0 mc
    else if k >= n then loop i (j + 1) 0 0 (max c mc)
    else let hi = i / w in
         let wi = i mod w in
         let hj = j / w in
         let wj = j mod w in
         let hk = k / w in
         let wk = k mod w in
         if a.(hi).[wi] = '.' &&
              a.(hj).[wj] = '.' &&
                a.(hk).[wk] = '.' &&
                  (dist (hi, wi) (hk, wk) <= d || dist (hj, wj) (hk, wk) <= d) then loop i j (k + 1) (c + 1) mc
         else loop i j (k + 1) c mc in
  loop 0 1 0 0 0

let () =
  scanf "%d %d %d " (fun h w d ->
    Array.init h (fun _ -> scanf "%s " @@ fun x -> x) |> solve h w d) |> printf "%d\n"
